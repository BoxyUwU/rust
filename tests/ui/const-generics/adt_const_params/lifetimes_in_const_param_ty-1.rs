#![feature(adt_const_params, generic_const_exprs)]
#![allow(incomplete_features)]

// Regression test for #125564 where we would wind up having erased lifetimes in type system
// constants in MIR without them ever being replaced with region variables during mir typeck.

const fn concat_strs() -> &'static str {
    //~^ ERROR: mismatched types
    const fn concat_arr<const M: usize, const N: usize>(a: [u8; M], b: [u8; N]) -> [u8; M + N] {}
    //~^ ERROR: mismatched types

    impl<const A: &'static str, const B: &'static str> Inner<A, B>
    //~^ ERROR: cannot find type `Inner` in this scope
    //~| ERROR: `&'static str` can't be used as a const parameter type
    //~| ERROR: `&'static str` can't be used as a const parameter type
    where
        [(); A.len()]:,
        [(); B.len()]:,
        [(); A.len() + B.len()]:,
    {
        const ABSTR: &'static str = unsafe {
            std::str::from_utf8_unchecked(&concat_arr(
                A.as_ptr().cast().read(),
                //~^ WARN: type annotations needed
                //~| WARN: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2018!
                B.as_ptr().cast().read(),
                //~^ WARN: type annotations needed
                //~| WARN: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2018!
            ))
        };
    }
}

fn main() {}
