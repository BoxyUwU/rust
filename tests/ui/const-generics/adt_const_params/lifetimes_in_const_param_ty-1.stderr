error[E0412]: cannot find type `Inner` in this scope
  --> $DIR/lifetimes_in_const_param_ty-1.rs:9:56
   |
LL |     impl<const A: &'static str, const B: &'static str> Inner<A, B>
   |                                                        ^^^^^ not found in this scope

error[E0741]: `&'static str` can't be used as a const parameter type
  --> $DIR/lifetimes_in_const_param_ty-1.rs:9:19
   |
LL |     impl<const A: &'static str, const B: &'static str> Inner<A, B>
   |                   ^^^^^^^^^^^^

error[E0741]: `&'static str` can't be used as a const parameter type
  --> $DIR/lifetimes_in_const_param_ty-1.rs:9:42
   |
LL |     impl<const A: &'static str, const B: &'static str> Inner<A, B>
   |                                          ^^^^^^^^^^^^

error[E0308]: mismatched types
  --> $DIR/lifetimes_in_const_param_ty-1.rs:4:27
   |
LL | const fn concat_strs() -> &'static str {
   |          -----------      ^^^^^^^^^^^^ expected `&str`, found `()`
   |          |
   |          implicitly returns `()` as its body has no tail or `return` expression

error[E0308]: mismatched types
  --> $DIR/lifetimes_in_const_param_ty-1.rs:6:84
   |
LL |     const fn concat_arr<const M: usize, const N: usize>(a: [u8; M], b: [u8; N]) -> [u8; M + N] {}
   |              ----------                                                            ^^^^^^^^^^^ expected `[u8; M + N]`, found `()`
   |              |
   |              implicitly returns `()` as its body has no tail or `return` expression
   |
note: consider returning one of these bindings
  --> $DIR/lifetimes_in_const_param_ty-1.rs:6:57
   |
LL |     const fn concat_arr<const M: usize, const N: usize>(a: [u8; M], b: [u8; N]) -> [u8; M + N] {}
   |                                                         ^           ^

warning: type annotations needed
  --> $DIR/lifetimes_in_const_param_ty-1.rs:20:35
   |
LL |                 A.as_ptr().cast().read(),
   |                                   ^^^^
   |
   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2018!
   = note: for more information, see issue #46906 <https://github.com/rust-lang/rust/issues/46906>
   = note: `#[warn(tyvar_behind_raw_pointer)]` on by default

warning: type annotations needed
  --> $DIR/lifetimes_in_const_param_ty-1.rs:23:35
   |
LL |                 B.as_ptr().cast().read(),
   |                                   ^^^^
   |
   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2018!
   = note: for more information, see issue #46906 <https://github.com/rust-lang/rust/issues/46906>

error: aborting due to 5 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0308, E0412, E0741.
For more information about an error, try `rustc --explain E0308`.
